ext/LuxEnzymeExt.jl

```diff
@@ -0,0 +1,26 @@
+module LuxEnzymeExt
+
+using ADTypes: AutoEnzyme
+using Enzyme: Enzyme
+using Lux: Lux
+using Setfield: @set!
+
+function Lux.Experimental.compute_gradients(::AutoEnzyme, objective_function::F, data,
+        ts::Lux.Experimental.TrainState) where {F}
+    dps = Enzyme.make_zero(ts.parameters)
+    fwd, rev = Enzyme.autodiff_thunk(
+        Enzyme.ReverseSplitWithPrimal, Enzyme.Const{typeof(objective_function)},
+        Enzyme.Active, Enzyme.Const{typeof(ts.model)},
+        Enzyme.Duplicated{typeof(ts.parameters)},
+        Enzyme.Const{typeof(ts.states)}, Enzyme.Const{typeof(data)})
+    tape, (loss, st_new, stats), shadow_result = fwd(
+        Enzyme.Const(objective_function), Enzyme.Const(ts.model),
+        Enzyme.Duplicated(ts.parameters, dps), Enzyme.Const(ts.states), Enzyme.Const(data))
+    rev(Enzyme.Const(objective_function), Enzyme.Const(ts.model),
+        Enzyme.Duplicated(ts.parameters, dps), Enzyme.Const(ts.states), Enzyme.Const(data),
+        (one(loss), Enzyme.make_zero(st_new), Enzyme.make_zero(stats)), tape)
```

Comment:
If you're just computing the derivatives directly here it is more efficient to just use autodiff rather than autodiff_thunk (which may be slower for requiring things to be preserved from fwd pass to a later rev pass and assumes data may be overwritten from fwd to rev)

Reply:
It also may be cool to extend the API to now for dps to be re-allocated every training step. You can re-use the same dps allocation for the gradient from iteration to iteration (as long as it's zero'd in place before each gradient call)

Reply:
> If you're just computing the derivatives directly here it is more efficient to just use autodiff rather than autodiff_thunk (which may be slower for requiring things to be preserved from fwd pass to a later rev pass and assumes data may be overwritten from fwd to rev)

the objective function returns `loss, st_new, stats`, so I need to seed the st_new and stats to have no derivative, how would I do that using the `autodiff` API?

Reply:
> It also may be cool to extend the API to now for dps to be re-allocated every training step. You can re-use the same dps allocation for the gradient from iteration to iteration (as long as it's zero'd in place before each gradient call)

Yes, I agree. We can probably cache the forward and reverse functions as well since they should stay constant.

I want to do that together with https://github.com/LuxDL/Lux.jl/issues/544

Reply:
> Yes, I agree. We can probably cache the forward and reverse functions as well since they should stay constant.

Actually, I think I can do it in this PR itself without any breaking change

Reply:
Is there a way to specify a structural zero instead of doing it like `Enzyme.make_zero(st_new)`?

Reply:
This is now done, first iteration creates the caches and we reuse it in subsequent iterations